### ISSUE TITLE
[BUG]  `test_fit_does_not_overwrite_hyper_params[FeatureUnion]` failing

### ISSUE THREAD

**Issue by lmmentel** [alan-turing-institute/sktime#1662]
Update: the failure has been silenced in the tests to enable refactor work on CI/CD, but the bug is still there.
To reproduce, the test should be run manually, or with the silencing disabled (`tests._config.EXCLUDED_TESTS`)

---

**Describe the bug**

In the refactored CI pipeline based on github actions #1620, and is blocking the PR.

The test `test_fit_does_not_overwrite_hyper_params[FeatureUnion]` from `tests/test_all_estimators.py` fails on linux with python3.6-3.9 and macos 3.6-3.9 with the error below.

Curiously the test are passing in CI pipelines currently on `main` branch.

```
____________ test_fit_does_not_overwrite_hyper_params[FeatureUnion] ____________
[gw0] darwin -- Python 3.7.12 /Users/runner/hostedtoolcache/Python/3.7.12/x64/bin/python

estimator_instance = FeatureUnion(n_jobs=None, preserve_dataframe=True,
             transformer_list=[('transformer1',
                   ...                                                              with_std=True)))],
             transformer_weights=None)

    def test_fit_does_not_overwrite_hyper_params(estimator_instance):
        """Check that we do not overwrite hyper-parameters in fit."""
        estimator = estimator_instance
        set_random_state(estimator)
    
        # Make a physical copy of the original estimator parameters before fitting.
        params = estimator.get_params()
        original_params = deepcopy(params)
    
        # Fit the model
        fit_args = _make_args(estimator, "fit")
        estimator.fit(*fit_args)
    
        # Compare the state of the model parameters with the original parameters
        new_params = estimator.get_params()
        for param_name, original_value in original_params.items():
            new_value = new_params[param_name]
    
            # We should never change or mutate the internal state of input
            # parameters by default. To check this we use the joblib.hash function
            # that introspects recursively any subobjects to compute a checksum.
            # The only exception to this rule of immutable constructor parameters
            # is possible RandomState instance but in this check we explicitly
            # fixed the random_state params recursively to be integer seeds.
>           assert joblib.hash(new_value) == joblib.hash(original_value), (
                "Estimator %s should not change or mutate "
                " the parameter %s from %s to %s during fit."
                % (estimator.__class__.__name__, param_name, original_value, new_value)
            )
E           AssertionError: Estimator FeatureUnion should not change or mutate  the parameter transformer_list from [('transformer1', SeriesToSeriesRowTransformer(check_transformer=False,
E                                          transformer=StandardScaler(copy=True,
E                                                                     with_mean=True,
E                                                                     with_std=True))), ('transformer2', SeriesToSeriesRowTransformer(check_transformer=False,
E                                          transformer=StandardScaler(copy=True,
E                                                                     with_mean=True,
E                                                                     with_std=True)))] to [('transformer1', SeriesToSeriesRowTransformer(check_transformer=False,
E                                          transformer=StandardScaler(copy=True,
E                                                                     with_mean=True,
E                                                                     with_std=True))), ('transformer2', SeriesToSeriesRowTransformer(check_transformer=False,
E                                          transformer=StandardScaler(copy=True,
E                                                                     with_mean=True,
E                                                                     with_std=True)))] during fit.
E           assert '7f94d1fc7e1f...888be251ce7b2' == 'b03f493febd2...c60681b4af6e4'
E             - b03f493febd2f1d6da1c60681b4af6e4
E             + 7f94d1fc7e1f285e1e5888be251ce7b2

estimator  = FeatureUnion(n_jobs=None, preserve_dataframe=True,
             transformer_list=[('transformer1',
                   ...                                                              with_std=True)))],
             transformer_weights=None)
estimator_instance = FeatureUnion(n_jobs=None, preserve_dataframe=True,
             transformer_list=[('transformer1',
                   ...                                                              with_std=True)))],
             transformer_weights=None)
fit_args   = (                                                var_0
0   0    -0.116020
1     0.343339
2    -0.464066
3...
1   0    ...0
7     1
8     0
9     0
10    0
11    0
12    1
13    1
14    1
15    1
16    0
17    0
18    1
19    1
dtype: int64)
new_params = {'n_jobs': None, 'preserve_dataframe': True, 'transformer1': SeriesToSeriesRowTransformer(check_transformer=False,
   ...                                                        with_std=True)), 'transformer1__check_transformer': False, ...}
new_value  = [('transformer1', SeriesToSeriesRowTransformer(check_transformer=False,
                             transformer=Stand...                              with_mean=True,
                                                        with_std=True)))]
original_params = {'n_jobs': None, 'preserve_dataframe': True, 'transformer1': SeriesToSeriesRowTransformer(check_transformer=False,
   ...                                                        with_std=True)), 'transformer1__check_transformer': False, ...}
original_value = [('transformer1', SeriesToSeriesRowTransformer(check_transformer=False,
                             transformer=Stand...                              with_mean=True,
                                                        with_std=True)))]
param_name = 'transformer_list'
params     = {'n_jobs': None, 'preserve_dataframe': True, 'transformer1': SeriesToSeriesRowTransformer(check_transformer=False,
   ...                                                        with_std=True)), 'transformer1__check_transformer': False, ...}
```

**To Reproduce**

Run the test with:

```bash
pytest sktime/tests/test_all_estimators.py
```

**Expected behavior**

Test passes

**Additional context**

**Versions**

See github actions under #1620 

</details>

**Comment by fkiraly** [comment]
I think the bug is straightforward, just looking at it (needs to be confirmed).

`FeatureUnion` has a `transformer` attribute which contains the components. Instead of fitting these directly where they are, the transformers should be cloned (using `deepcopy` or a safe `sklearn` `clone`) to a `transformer_` attribute and fitted there, without having side effects on the `transformer` attribute written in the init.

**Comment by fkiraly** [comment]
Relates to this section of the extension contract specification in the extension template which is probably violated:

```
        # any model parameters should be written to attributes ending in "_"
        #  attributes set by the constructor must not be overwritten
        #  if used, estimators should be cloned to attributes ending in "_"
        #  the clones, not the originals shoudld be used or fitted if needed
```

**Comment by fkiraly** [comment]
don't have time to fix it today, but if someone wants to pick it up quickly, I've attached the "good first issue" label

**Comment by lmmentel** [comment]
Thanks for the input @fkiraly! Let me try to make it work.

**Comment by lmmentel** [comment]
I found that `FeatureUnion` has a lot of code duplicated from [sklearn.pipeline](https://github.com/scikit-learn/scikit-learn/blob/0d378913b/sklearn/pipeline.py) so I got rid of that in #1665 and I'm getting a different test failing this time, namely: 

```
_____________________________________________ test_raises_not_fitted_error[FeatureUnion] ______________________________________________

estimator_instance = FeatureUnion(n_jobs=None, preserve_dataframe=True,
             transformer_list=[('transformer1',
                   ...                                                              with_std=True)))],
             transformer_weights=None)

    def test_raises_not_fitted_error(estimator_instance):
        """Check that we raise appropriate error for unfitted estimators."""
        estimator = estimator_instance
        # call methods without prior fitting and check that they raise our
        # NotFittedError
        for method in NON_STATE_CHANGING_METHODS:
            if hasattr(estimator, method):
                args = _make_args(estimator, method)
                with pytest.raises(NotFittedError, match=r"has not been fitted"):
>                   getattr(estimator, method)(*args)
E                   Failed: DID NOT RAISE <class 'sktime.exceptions.NotFittedError'>

sktime/tests/test_all_estimators.py:357: Failed
```

Which is a bit weird since if I type the minimal example it works as expected:

```python
from sktime.series_as_features.compose import FeatureUnion
fu = FeatureUnion([("scaler", MinMaxScaler())])
fu.transform([1])
...
    raise NotFittedError(msg % {'name': type(estimator).__name__})
sklearn.exceptions.NotFittedError: This MinMaxScaler instance is not fitted yet. Call 'fit' with appropriate arguments before using this estimator.
```

The test though. I think it's because it's a sklearn exception and not sktime's and the message is slightly different.

**Comment by lmmentel** [comment]
On a broader note, I don't know if `FeatureUnion` is well placed being tested on par with other estimators since it's a "meta" estimator that takes other estimators as arguments. From that perspective it's closer to a `Pipeline` object. Any thoughts @fkiraly @mloning ?

**Comment by fkiraly** [comment]
>  I think it's because it's a sklearn exception and not sktime's and the message is slightly different.

Yes, the test looks for the text `has not been fitted`, but the expectation raised has `is not fitted` instead. So either changing the error message or the text that the test is looking for should fix this, would suggest the former if that's easy.

**Comment by fkiraly** [comment]
> On a broader note, I don't know if FeatureUnion is well placed being tested on par with other estimators since it's a "meta" estimator that takes other estimators as arguments. From that perspective it's closer to a Pipeline object. Any thoughts @fkiraly @mloning ?

That's perfectly fine, since composites (what you call "meta" estimators) usually follow the template of an estimator type, in this case a transformer. That's the same with `sklearn` pipeline or grid search, which results in an object that behaves just like your normal supervised learner or transformer.

**Comment by mloning** [comment]
We have our own `sktime.exceptions.NotFittedError`, seems like the FeatureUnion is raising scikit-learn's one.

**Comment by lmmentel** [comment]
It seems this should be reopened since the test is not passing still, see https://github.com/lmmentel/sktime/runs/4498946788?check_suite_focus=true As you can see it's failing for all linux host and for macOS with py37.

I've tried some additional ideas but so far I'm not able to make it work. Could someone who knows more about `PaneltoPanelTransformer` and `SeriesToSeriesRowTransformer` look into this. I think the issue might be related to those transformers not being "good"  meta transformers. @mloning @fkiraly

**Comment by fkiraly** [comment]
Why is this closed, @TonyBagnall? It doesn't seem resolved (on the specific combination of py version and OS), we need this to pass on all combinations.

**Comment by TonyBagnall** [comment]
did I close it? Sorry, must have been a mistake on my part!

**Comment by lmmentel** [comment]
This can be closed now since the failing test is now being excluded through #1792.

**Comment by fkiraly** [comment]
no, not really, @lmmentel!

Because this is still a bug! We're just silencing it temporarily, until someone fixes it...

**Comment by lmmentel** [comment]
OK, than I guess the title and description should be updated to reflect the actual issue and how to reproduce.

**Comment by fkiraly** [comment]
sure - updated the text of the bug description. I think the title is still accurate?

**Comment by fkiraly** [comment]
Fixed by #1810
